### DISCUSSION TITLE
Can't access theme from styled component and question about sx props vs style props

### DISCUSSION THREAD

**Discussion by coreybruyere** [system-ui/theme-ui#1567] [General]
I have a React TS-based design system built with emotion, styled-system, and theme-ui and am encountering a TS error I can't seem to resolve. It seems like a simple fix, I just haven't been able to debug it.

I found a work around by just using `bg`, so utilizing styled function props, and I plan on utilizing the `sx` prop for all other styling needs.

The error is:
`Property 'colors' does not exist on type 'Theme'.`

I'm trying to access `theme` in a styled component with:

```
background-color: ${({ theme }) => theme.colors.border};
```

My question is: Why isn't it possible to style utilizing the theme prop inside of a styled component?

**Comment by hasparus** [comment]
Could you share your whole component?
Is colors present on theme at runtime and the error is purely in types world, or is colors also missing at runtime? Do you import "styled" from styled-components or @emotion/styled?

**Reply by coreybruyere** [reply]
I was trying to reference color as noted above in the `Divider` component inside of `packages/core/Divider` in this project: https://github.com/coreybruyere/skeleton-ui

**Reply by hasparus** [reply]
Oh, okay, I get it.

I made a small CodeSandbox with a possible solution for you. Give it a try.

https://codesandbox.io/s/theme-ui-emotion-styled-component-theme-5gq6e

We could do that `declare module` on our side, but we don't know your exact theme on library level, so the only thing we can do is informing Emotion _"hey this is a general Theme-UI-style theme"_, and your styled components will need a lot of optional chaining `?.`

I'm using the trick from CodeSandbox in my personal projects quite a lot — it should work for any Theme UI and Emotion version.

**Reply by coreybruyere** [reply]
Thanks @hasparus - I tried replicating your Codesandbox code inside of my ThemeProvider component, theme, and Divider component and am still seeing the error? Am I declaring the theme interface module correctly?

I'm linking directly to the files in question, any guidance would be greatly appreciated!

https://github.com/coreybruyere/skeleton-ui/tree/master/packages/core/src/components/ThemeProvider
https://github.com/coreybruyere/skeleton-ui/tree/master/packages/core/src/components/Divider
https://github.com/coreybruyere/skeleton-ui/blob/master/packages/core/src/theme.ts

**Reply by hasparus** [reply]
Hey mate, I made a PR to your repo — https://github.com/coreybruyere/skeleton-ui/pull/12.

Next time, try to look for "references" and "implementations" thingy in your IDE. These numbers should be higher than zero when you're declaration merging the thing you want :D

![image](https://user-images.githubusercontent.com/15332326/111374574-90aa4f00-869d-11eb-8726-c61c3c1c652e.png)

**Reply by hasparus** [reply]
Quick translation to English to explain what's happening with declaration merging and module augmentation for people who come here from Google.

1. `declare module "some-package"` — Let's pretend we're in "some-package".
2. `export interface SomeType extends MyType` — Remember SomeType? It was already exported from here. Let's make it extend MyType now.

**Reply by coreybruyere** [reply]
Thank you for the help! That worked. 

Just one last question while I have your attention:

Is the purpose of the `sx` prop to replace style props from `styled-system`? If so, in your experience, what's the best way to _consume_ and _build_ on top of a theme-ui-based library? I noticed when consuming a component with `sx` props and applying additional styles on top of said component via `sx` prop that all previous styles from the original component are wiped out.

In the library you made a PR in - you might have noticed I'm using styled-system. I'm trying to make a consistent API from a consumer's perspective, meaning that I'd like to either solely stick with the `sx` prop or stick with style props with the option to easily import and extend a components styles with the `styled` API using regular css-in-js. I'm just trying to wrap my head around which direction to head in, in terms of API.

Thanks for taking care of this repo and thanks again for the help.

**Reply by hasparus** [reply]
> I noticed when consuming a component with sx props and applying additional styles on top of said component via sx prop that all previous styles from the original component are wiped out.

If you're using a JSX pragma, merging `sx` props should work out of the box — `sx` is compiled to `className` so then you have `sx`, and `className` from props.

If you're not transpiling your JSX, you'll have to merge `sx` objects.

```tsx
sx={{
  color: 'blue',
   ...props.sx,
}}
```

This is similar to merging classNames when using pragma / classic CSS.

```tsx
className={join("Button", active && "Button--active", props.className)}
```

So, here's a question — how opinionated is your components library? Is it a reusable pack of components to be styled in whatever way the user is accustomed to? Then, accepting styles from props through either `sx`, `className` or `style` props seems like a nice permissive API.

However, if you're building a closed design language, you'd probably accept neither of these props, maybe leaving some escape hatch like seek-oss/braid-design-system does with Box.

I'm using Chakra in one commercial project, and I've grown... skeptical of style props, but that's just my personal opinion. I'm familiar with a few quite ambitious designers and I'd rather have an access to a superset of CSS (like sx) than a subset or an equivalent set of JSX props. There's always a (far albeit realistic) possibility of static optimization of some of your `sx` props and transforming them to classNames at build time.

**Reply by coreybruyere** [reply]
@hasparus - I tried merging sx objects after adding a jsx pragma and am not getting the result I'd expect. Here's a [Pull request](https://github.com/coreybruyere/skeleton-ui/pull/13/files#diff-cadfce3473f7dc978c8133d96fe30f5c7231ca05e9b435a7f912486b1046d42d) with changes.

Does anything look wrong here? 

I'd expect any additional styled props **or** sx props I apply on top of an existing component (styled with styled props or sx prop) to override the styles that were previously defined. In [this case](https://github.com/coreybruyere/skeleton-ui/pull/13/files#diff-0d59eb9ff7490c7f4df0fad40c208f36c77208e5bb640fabe9d8615cecc89bc1) I expect a `p` value of `5` with a `fontSize` of `40`:

```jsx
<Tag p={5} sx={{fontSize: 40}}>This is a Tag</Tag>
```

For context: I'm essentially trying to create a small library that supports both styled-system styled props AND theme-ui `sx` prop.

**Reply by hasparus** [reply]
In your PR `py` and `px` overwrite `p`, right?

> For context: I'm essentially trying to create a small library that supports both styled-system styled props AND theme-ui sx prop.

I'm afraid you'll have to manage the "priority" manually. I'd expect that in your example all styles you wrote are emitted but the class is not in ordered as you'd like. Is it true?

**Reply by coreybruyere** [reply]
Yes this is true, `py` and `px` are overriding `p`. How would I go about managing that? To ensure amended props override preexisting props?

**Reply by hasparus** [reply]
I, personally, stray from style props because they give you a second way to do the same thing for a slightly more concise syntax.

If you want to say _"whenever `p` is passed from top-level, I want it to override my `px` and `py`"_ you could write it like this:

```tsx
const paddingFromProps = props.p || props.padding;
const padding = paddingFromProps ? { p: paddingFromProps } : { px: 1, py: 2 }

<div 
  sx={{
    ...padding,
    ...sx,
  }}
/>
```
